### PR TITLE
Laravel 6.x

### PR DESCRIPTION
**Pull request by Jellyfrog**
This pull request includes the changes for upgrading to Laravel 6.x. Feel free to commit any additional changes to the `shift-26000` branch.

**Before merging**, you need to:

- Checkout the `shift-26000` branch
- Review **all** pull request comments for additional changes
- Update your dependencies for Laravel 6
- Run `composer update` (if the scripts fail, add `--no-scripts`)
- Thoroughly test your application ([no tests?](https://confidentlaravel.com))

If you need help with your upgrade, check out the [Human Shifts](https://laravelshift.com/human-shifts). You may also join the [Shifty Coders](https://laravelshift.com/shifty-coders) Slack workspace to level-up your Laravel skills.

### CODE REVIEW

**Comment by Jellyfrog** [comment]
:x: Shift could not upgrade the following files since they differed from the default Laravel version. You will need to compare these files against the [default Laravel 6 versions](https://github.com/laravel-shift/laravel-6.0/) and merge any changes:

- [ ] app/Http/Kernel.php
- [ ] resources/lang/en/validation.php

**Comment by Jellyfrog** [comment]
:warning: Laravel 6.x consolidated references to `/home` as a constant within the `RouteServiceProvider`. This makes it easier to maintain and change in a single place.

Shift [automated this change](https://github.com/librenms/librenms/pull/11397/commits/334ec857e833341ca03f29dc94904261c1ce48ae?w=1), but could not upgrade the following files since they differed from the default Laravel version. You should compare these files against the [default Laravel 6 versions](https://github.com/laravel-shift/laravel-6.0/) and merge this change:

- [ ] app/Http/Controllers/Auth/LoginController.php

**Comment by Jellyfrog** [comment]
:warning: Since you are upgrading from an older version of Laravel, Shift defaulted your customized configuration files to make future Shifts smoother. This was done in [a separate commit](https://github.com/librenms/librenms/pull/11397/commits/3a5a46d1d9a274b6568590f7404f27bc4763fa9f?w=1) so you may easily review and backfill your customizations once you are running the latest version.

- [ ] config/auth.php
- [ ] config/broadcasting.php
- [ ] config/cache.php
- [ ] config/database.php
- [ ] config/filesystems.php
- [ ] config/logging.php
- [ ] config/mail.php
- [ ] config/queue.php
- [ ] config/services.php

**Comment by Jellyfrog** [comment]
:information_source: Laravel 6 changed the default Redis client from `predis` to `phpredis`. You may keep using `predis` by setting `REDIS_CLIENT=predis` for your environment.

However, if possible, consider switching to [phpredis](https://github.com/phpredis/phpredis) to gain the performance of its PHP extension and avoid using the deprecated `predis` dependency which will be removed in Laravel 7.0.

**Comment by Jellyfrog** [comment]
:information_source: Shift [updated your dependencies](https://github.com/librenms/librenms/pull/11397/commits/632fb2f753c26f76267da8c6bff6d2c0cf676400?w=1) for Laravel 6. While many of the popular packages are reviewed, you may have to update additional packages in order for your application to be compatible with Laravel 6.

Watch [dealing with dependencies](https://laravelshift.com/videos/update-incompatible-composer-dependencies) for tips on handling package incompatibilities.

**Comment by Jellyfrog** [comment]
:x: Laravel 6 added the `viewAny` method to authorization policies attached to controllers. You will need to add the following method to your policies and update the logic. Otherwise calls your controller's `index` action will be rejected.

```php
/**
 * Determine whether the user can view any model resource.
 *
 * @param  \App\User $user
 * @return mixed
 */
public function viewAny(User $user)
{
    return true;
}
```

**Comment by Jellyfrog** [comment]
:warning: In Laravel 6, the `update` method of the `BelongsTo` relationship no longer provides mass assignment protection or fires Eloquent events. If you wish to preserve this behavior, you should call `update` on the model itself.

```php
// no mass assignment protection or events...
$post->user()->update(['foo' => 'bar']);

// mass assignment protection and events...
$post->user->update(['foo' => 'bar']);
```

Shift found potential uses of the `update` method in:

- [ ] app/Models/PollerGroup.php

**Comment by Jellyfrog** [comment]
:information_source: The Eloquent model's `toArray()` method will now cast any attributes that implement `Illuminate\Contracts\Support\Arrayable` to an array. While unlikely to impact your application, this change could effect code which relies upon the previous data type or output.

**Comment by Jellyfrog** [comment]
:information_source: Laravel 6 now requires Carbon 2. While Shift reviewed your application for common breaking changes, you may want to review the [Carbon 2 migration notes](https://carbon.nesbot.com/docs/#api-carbon-2) for additional changes.

**Comment by Jellyfrog** [comment]
:warning: Previous versions of Laravel would bind parameters passed to the URL helper methods as placeholder values even if the names mismatched. In Laravel 6, named parameters not matching placeholders are attached to the query string instead.

Shift detected the following `route()` and `action()` calls passing named parameters you should review to ensure they match the route placeholder name.

- [ ] resources/views/user/edit.blade.php

**Comment by Jellyfrog** [comment]
:information_source: Laravel 6 made [performance optimizations](https://github.com/laravel/framework/pull/28153) for _integer_ key types. If you are using a _string_ as your model's primary key, you may set the `$keyType` property on your model.

```php
/**
 * The "type" of the primary key ID.
 *
 * @var string
 */
protected $keyType = 'string';
```

**Comment by Jellyfrog** [comment]
:information_source: The `mandrill` and `sparkpost` mail drivers, as well as the `rackspace` storage driver were removed in Laravel 6. If you were using these drivers, you may adopt a community maintained package which provides the driver.

**Comment by Jellyfrog** [comment]
:information_source: Previous versions of Laravel would retry jobs indefinitely. Beginning with Laravel 6, the `php artisan queue:work` now tries a job one time by default. If you want to force jobs to be tried indefinitely, you may pass the `--tries=0` option.

**Comment by Jellyfrog** [comment]
:information_source: Shift noticed you have additional namespaces in your application. You should consider using the [Consolidate Namespaces Shift](https://laravelshift.com/laravel-consolidate-custom-namespaces) to simplify your namespaces into the default Laravel `App` namespace.

**Comment by Jellyfrog** [comment]
> ℹ️ Laravel 6 made [performance optimizations](https://github.com/laravel/framework/pull/28153) for _integer_ key types. If you are using a _string_ as your model's primary key, you may set the `$keyType` property on your model.
> 
> ```
> /**
>  * The "type" of the primary key ID.
>  *
>  * @var string
>  */
> protected $keyType = 'string';
> ```
Do we have string primary keys anywhere?

**Comment by Jellyfrog** [comment]
Also there is something weird with phpunit... It just cuts the output random sometimes, these should all give the same output:
https://travis-ci.com/github/librenms/librenms/jobs/319688676
https://travis-ci.com/github/librenms/librenms/jobs/319688680
https://travis-ci.com/github/librenms/librenms/jobs/319688755
https://travis-ci.com/github/librenms/librenms/jobs/319688757

**Comment by Jellyfrog** [comment]
I'm also guessing its a bug I needed to patch the re-generated tests, they work locally but fails on travis: 0e6ae7208ae821221737b9746f6ebb948d119831

**Comment by Jellyfrog** [comment]
- [ ]  fix/replace `wpb/string-blade-compiler`

File: config/auth.php [hunk 1]
```diff
@@ -105,7 +105,21 @@
             'provider' => 'users',
             'table' => 'password_resets',
             'expire' => 60,
+            'throttle' => 60,
```
Comment (murrant) [inline config/auth.php:108]: This needs to be removed/unset

File: composer.json [hunk 2]
```diff
@@ -32,41 +39,41 @@
...
         "pear/console_table": "^1.3",
         "php-amqplib/php-amqplib": "^2.0",
         "phpmailer/phpmailer": "~6.0",
         "rmccue/requests": "^1.7",
-        "spatie/laravel-cors": "1.3.*",
+        "spatie/laravel-cors": "^1.6",
         "symfony/yaml": "^4.0",
         "tecnickcom/tcpdf": "~6.2.0",
-        "tightenco/ziggy": "^0.7.1",
-        "wpb/string-blade-compiler": "3.8.x-dev",
```
Comment (murrant) [inline composer.json:53]: string-blade-compiler was removed, re-add it, but looks like upstream has been updated with Laravel 6.0 support.  https://github.com/TerrePorter/StringBladeCompiler/commit/d067645a2601be8256ae758d1612e8fda197cc56

File: app/Models/PollerGroup.php [hunk 3]
```diff
@@ -43,7 +43,7 @@ public static function boot()
         static::deleting(function (PollerGroup $pollergroup) {
             // handle device pollergroup fallback to default poller
             $default_poller_id = \LibreNMS\Config::get('distributed_poller_group');
-            $pollergroup->devices()->update(['poller_group' => $default_poller_id]);
+            $pollergroup->devices->update(['poller_group' => $default_poller_id]);
```
Comment (murrant) [inline app/Models/PollerGroup.php:46]: This is probably fine, but could be very slow.

Comment (Jellyfrog) [reply app/Models/PollerGroup.php:46]: The reason is stated in f09a7671703fc9090f439eb13c1d8c538184953e, what do you say?

File: config/auth.php (hunk 1 above)
Comment (Jellyfrog) [reply config/auth.php:108]: Imo we should keep it, seems like a sane enough setting

File: composer.json (hunk 2 above)
Comment (Jellyfrog) [reply composer.json:53]: When adding it with `FORCE=1 composer require wpb/string-blade-compiler="4.0.x-dev"`
I get;
```
> Illuminate\Foundation\ComposerScripts::postAutoloadDump
> @php artisan package:discover --ansi

   Illuminate\Contracts\Container\BindingResolutionException  : Target class [blade.compiler] does not exist.

  at /Users/jellyfrog/code/librenms/vendor/laravel/framework/src/Illuminate/Container/Container.php:805
    801|
    802|         try {
    803|             $reflector = new ReflectionClass($concrete);
    804|         } catch (ReflectionException $e) {
  > 805|             throw new BindingResolutionException("Target class [$concrete] does not exist.", 0, $e);
    806|         }
    807|
    808|         // If the type is not instantiable, the developer is attempting to resolve
    809|         // an abstract type such as an Interface or Abstract Class and there is

  Exception trace:

  1   Illuminate\Foundation\Application::Illuminate\Foundation\{closure}(Object(DarkGhostHunter\Larapoke\LarapokeServiceProvider))
      [internal]:0

  2   ReflectionException::("Class blade.compiler does not exist")
      /Users/jellyfrog/code/librenms/vendor/laravel/framework/src/Illuminate/Container/Container.php:803

  Please use the argument -v to see more details.
```

Note that It talks about LarapokeServiceProvider, but I'm guessing thats a side effect.

Comment (Jellyfrog) [reply composer.json:53]: This was the culpit, I removed it for now... https://github.com/librenms/librenms/pull/8907.  
@murrant thoughts?

File: app/Models/PollerGroup.php (hunk 3 above)
Comment (murrant) [reply app/Models/PollerGroup.php:46]: I say we don't need those events right now.  I would rather skip them.

File: composer.json (hunk 2 above)
Comment (murrant) [reply composer.json:53]: It now uses auto-registration.  The manual binding isn't required.  That is the correct fix.

File: config/auth.php (hunk 1 above)
Comment (murrant) [reply config/auth.php:108]: Except, it was set before and we had to remove it :)

Comment (murrant) [reply config/auth.php:108]: Hmm, git says it wasn't.   We can leave it and see if there are any complaints

Comment (Jellyfrog) [reply config/auth.php:108]: git doesn't lie!

File: app/Models/PollerGroup.php (hunk 3 above)
Comment (Jellyfrog) [reply app/Models/PollerGroup.php:46]: Fixed

File: LibreNMS/Validations/Php.php [hunk 4]
```diff
@@ -48,11 +48,11 @@ public function validate(Validator $validator)
 
     private function checkVersion(Validator$validator)
     {
-        $min_version = '5.6.4';
+        $min_version = '7.2';
 
         // if update is not set to false and version is min or newer
         if (Config::get('update') && version_compare(PHP_VERSION, $min_version, '<')) {
-            $validator->warn('PHP version 5.6.4 is the minimum supported version as of January 10, 2018.  We recommend you update to PHP a supported version of PHP (7.1 suggested) to continue to receive updates.  If you do not update PHP, LibreNMS will continue to function but stop receiving bug fixes and updates.');
+            $validator->warn('PHP version 7.2 is the minimum supported version as of January 10, 2018.  We recommend you update to PHP a supported version of PHP (7.3 suggested) to continue to receive updates.  If you do not update PHP, LibreNMS will continue to function but stop receiving bug fixes and updates.');
```
Comment (Jellyfrog) [inline LibreNMS/Validations/Php.php:55]: @murrant this is obviously wrong, but I'm not sure what it should say, we check for php version in like 4 files, its rather confusing.  
It should probably be another PR that we merge before, that warns about new requirements.

**Comment by Jellyfrog** [comment]
This depends on https://github.com/librenms/librenms/pull/11465, or at-least `php artisan view:clear`

**Comment by Jellyfrog** [comment]
Kinda depends on https://github.com/librenms/librenms/pull/11470, and also updating `daily.sh` with minimum php version

**Comment by Jellyfrog** [comment]
Blocked by:

-  #11465
- #11526 
- #11531

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-64-release-changelog-may-2020/12290/1

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/users-get-redirected-to-login-page/12424/2
